File: C-Plus-Plus/Bit--Manipulation/Swap_even_and_odd_bits.cpp
//Swap Even and Odd Bits

#include <bits/stdc++.h>
using namespace std;

int main()
{
	unsigned int x;
	cout << "Enter number \n";
	cin >> x;

	// Get all even bits of x
	unsigned int even_bits = x & 0xAAAAAAAA;

	// Get all odd bits of x
	unsigned int odd_bits = x & 0x55555555;
	// Right shift even bits
	even_bits >>= 1;
	// Left shift odd bits
	odd_bits <<= 1;
	// Combine even and odd bits
	cout << (even_bits | odd_bits);
}

/*
Time Complexity: O(1)
Space Complexity: O(1)
*/

/*Sample Input:
Example 1:- 
Enter number
50
Example 2:-
Enter number
994

Sample Output:
Example 1:- 
49
Example 2:-
997
*/

File: Java/other/SwapEvenAndOddBits.java
/* This program will swap even and odd bits of
 the given integer.
 For Ex: 
         If the given number is 21(00010101) the ouput 
         should be 42(00101010).
 */
 import java.util.*;
 import java.lang.*;
 import java.io.*;

 class swap_even_odd_bits
 {
 	public static void main (String[] args) throws java.lang.Exception
 	{

        Scanner scn = new Scanner(System.in);
         System.out.println("Enter the number:");
         int n=scn.nextInt();

        //e represents hexadecimal number 1010(10)
         int e=0xAAAAAAAA ;

        //o represents hexadecimal number 0101(5)
         int o=0x55555555;

        //this will preserve all even bits of n and
        //make all odd bits of n 0.
         int even= e&n;

         //this will preserve all odd bits of n and
        //make all even bits of n 0.
         int odd= o&n;

         even >>= 1;
         odd <<= 1;

         System.out.println(even | odd); 
 	}
 }

 /*           TEST CASES

    1)Input:
            Enter the number:
            25
     Output:
            38
    
    2)Input:
            Enter the number:
            32
     Output:
            16
            

     Time Complexity:  O(1)
     Space Complexity: O(1)
 */
